tests/integration/test_array_comprehension.py

```diff
@@ -0,0 +1,79 @@
+import pytest
+
+from guppylang.decorator import guppy
+from guppylang.module import GuppyModule
+from guppylang.prelude.builtins import array
+from guppylang.prelude.quantum import qubit
+
+import guppylang.prelude.quantum_functional as quantum
+from tests.util import compile_guppy
+
+
+def test_basic(validate):
+    @compile_guppy
+    def test() -> array[int, 10]:
+        return array(i + 1 for i in range(10))
+
+    validate(test)
+
+
+def test_basic_linear(validate):
+    module = GuppyModule("test")
+    module.load_all(quantum)
+    module.load(qubit)
+
+    @guppy(module)
+    def test() -> array[qubit, 42]:
+        return array(qubit() for _ in range(42))
+
+    validate(module.compile())
+
+
+def test_zero_length(validate):
+    @compile_guppy
+    def test() -> array[int, 0]:
+        return array(i for i in range(0))
+
+    validate(test)
+
+
+def test_capture(validate):
+    @compile_guppy
+    def test(x: int) -> array[int, 42]:
+        return array(i + x for i in range(42))
```

Comment:
Would be good to make this one an execution test to check it does actually capture the correct value of `x`....

Reply:
Agreed, but unfortunately it doesn't run yet until we have https://github.com/CQCL/hugr/issues/1627 and corresponding lowering in hugr-llvm.